{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a0c72ebf",
   "metadata": {},
   "source": [
    "# Scratch Pad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f14ad7b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages.\n",
    "import gzip\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8254e95",
   "metadata": {},
   "source": [
    "## EPO Ancestral Sequence vs Hg19 Reference Sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3661bafe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to load a epo ancestral sequence.\n",
    "def load_epo_seq(chrom):\n",
    "    \"\"\"\n",
    "    Returns the EPO ancestral sequence for a focal chromosome.\n",
    "    \"\"\"\n",
    "    # Intialize the ancestral sequence.\n",
    "    anc_seq = ''\n",
    "    # Open the fasta file.\n",
    "    with open(f'../data/homo_sapiens_ancestor_GRCh37_e71/homo_sapiens_ancestor_{chrom}.fa', 'r') as fa_data:\n",
    "        # Compile the ancestral sequence.\n",
    "        anc_seq = [line.strip().replace(' ', '') for line in fa_data if not line.startswith('>')]\n",
    "    return ''.join(anc_seq)\n",
    "\n",
    "# Define a function the load the hg19 reference sequence.\n",
    "def load_hg19_seq(chrom):\n",
    "    \"\"\"\n",
    "    Returns the Hg19 reference sequence for a focal chromosome.\n",
    "    \"\"\"\n",
    "    # Intialize a flag for compiling the reference sequence.\n",
    "    compile_seq = False\n",
    "    # Intialize a list for storing reference sequence.\n",
    "    ref_seq = []\n",
    "    # Open the fasta file.\n",
    "    with gzip.open(f'../data/hg19/hg19.fa.gz', 'rt') as fa_data:\n",
    "        # For every line in the fasta file.\n",
    "        for line in fa_data:\n",
    "            # Clean the line.\n",
    "            line = line.strip()\n",
    "            # If this is a header line.\n",
    "            if line.startswith('>'):\n",
    "                # Check to see if this is the focal chromosome.\n",
    "                if line.lstrip('>') == f'chr{chrom}':\n",
    "                    # Start compiling the reference sequence.\n",
    "                    compile_seq = True\n",
    "                # Else-if this is the next chromosome's header line. \n",
    "                elif compile_seq:\n",
    "                    # Stop iterating after compiling the refrence sequence.\n",
    "                    break\n",
    "            # Else-if we are iterating through the focal chromosome.\n",
    "            elif compile_seq:\n",
    "                    # Update the reference sequence.\n",
    "                    ref_seq.append(line)\n",
    "    return ''.join(ref_seq)\n",
    "\n",
    "# Define a function to build a EPO alignment dictionary.\n",
    "def build_epo_aln_dicc(chrom):\n",
    "    \"\"\"\n",
    "    Build the EPO alignment dictionary for a specified chromosome.\n",
    "    \"\"\"\n",
    "    # Intialize a set of alleles to skip.\n",
    "    skip_set = {'N', '-', '.'}\n",
    "    # Intialize a dictionary to store the EPO alignment.\n",
    "    epo_aln_dicc = {}\n",
    "    # Load the Hg19 and EPO sequences.\n",
    "    hg19_seq = load_hg19_seq(chrom)\n",
    "    epo_seq = load_epo_seq(chrom)\n",
    "    \n",
    "    # Iterate through the chromosome.\n",
    "    for pos, (hg19, epo) in enumerate(zip(hg19_seq, epo_seq), start=1):\n",
    "        # If both alleles are valid.\n",
    "        if hg19 not in skip_set and epo not in skip_set:\n",
    "            # Update the dictionary.\n",
    "            epo_aln_dicc[pos] = (hg19.upper(), epo.upper())\n",
    "    return epo_aln_dicc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8acd0208",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chr1: Hg19 = 249250621bp; EPO = 249250621bp; QC'ed = True\n",
      "Chr2: Hg19 = 243199373bp; EPO = 243199373bp; QC'ed = True\n",
      "Chr3: Hg19 = 198022430bp; EPO = 198022430bp; QC'ed = True\n",
      "Chr4: Hg19 = 191154276bp; EPO = 191154276bp; QC'ed = True\n",
      "Chr5: Hg19 = 180915260bp; EPO = 180915260bp; QC'ed = True\n",
      "Chr6: Hg19 = 171115067bp; EPO = 171115067bp; QC'ed = True\n",
      "Chr7: Hg19 = 159138663bp; EPO = 159138663bp; QC'ed = True\n",
      "Chr8: Hg19 = 146364022bp; EPO = 146364022bp; QC'ed = True\n",
      "Chr9: Hg19 = 141213431bp; EPO = 141213431bp; QC'ed = True\n",
      "Chr10: Hg19 = 135534747bp; EPO = 135534747bp; QC'ed = True\n",
      "Chr11: Hg19 = 135006516bp; EPO = 135006516bp; QC'ed = True\n",
      "Chr12: Hg19 = 133851895bp; EPO = 133851895bp; QC'ed = True\n",
      "Chr13: Hg19 = 115169878bp; EPO = 115169878bp; QC'ed = True\n",
      "Chr14: Hg19 = 107349540bp; EPO = 107349540bp; QC'ed = True\n",
      "Chr15: Hg19 = 102531392bp; EPO = 102531392bp; QC'ed = True\n",
      "Chr16: Hg19 = 90354753bp; EPO = 90354753bp; QC'ed = True\n",
      "Chr17: Hg19 = 81195210bp; EPO = 81195210bp; QC'ed = True\n",
      "Chr18: Hg19 = 78077248bp; EPO = 78077248bp; QC'ed = True\n",
      "Chr19: Hg19 = 59128983bp; EPO = 59128983bp; QC'ed = True\n",
      "Chr20: Hg19 = 63025520bp; EPO = 63025520bp; QC'ed = True\n",
      "Chr21: Hg19 = 48129895bp; EPO = 48129895bp; QC'ed = True\n",
      "Chr22: Hg19 = 51304566bp; EPO = 51304566bp; QC'ed = True\n",
      "ChrX: Hg19 = 155270560bp; EPO = 155270560bp; QC'ed = True\n",
      "ChrY: Hg19 = 59373566bp; EPO = 59373566bp; QC'ed = True\n"
     ]
    }
   ],
   "source": [
    "# For every chromosome.\n",
    "for chrom in [i for i in range(1, 23)]+['X', 'Y']:\n",
    "    # Load the hg19 and ancestral sequence.\n",
    "    hg19_seq = load_hg19_seq(chrom)\n",
    "    epo_seq = load_epo_seq(chrom)\n",
    "    # Sanity check.\n",
    "    hg19_bp, epo_bp = len(hg19_seq), len(epo_seq)\n",
    "    print(f'Chr{chrom}: Hg19 = {hg19_bp}bp; EPO = {epo_bp}bp; QC\\'ed = {hg19_bp == epo_bp}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7af3dac",
   "metadata": {},
   "source": [
    "## Legacy Code vs New Code for Parsing a `.maf.gz` File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dd38e2f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Legacy Code From https://github.com/David-Peede/chimp_maf_to_vcf/chimp_maf_parser.py ###\n",
    "\n",
    "\n",
    "def maf_reader(maf_file):\n",
    "    \"\"\"\n",
    "    Reads in a .maf file and appends all lines in a synteny sequence alignment\n",
    "    block to a list which will be parsed by the function maf_parser.\n",
    "    \"\"\"\n",
    "    line = maf_file.readline()\n",
    "    while line[0] != \"a\":\n",
    "        line = maf_file.readline()\n",
    "    block = []\n",
    "    while line != \"\":\n",
    "        line = maf_file.readline()\n",
    "        if line == \"\" or line[0] == \"a\":\n",
    "            yield block\n",
    "            block = []\n",
    "        elif line[0] == \"s\": block.append(line)\n",
    "\n",
    "def maf_parser(maf_block):\n",
    "    \"\"\"\n",
    "    Takes the output from the maf_reader function and creates a dictionary for\n",
    "    each synteny sequence alignment block where the sequence name are the keys\n",
    "    and the values are a dictionary consisting of the sequence start, size,\n",
    "    strand oreintation, length, and the actual sequence.\n",
    "    \"\"\"\n",
    "    output = {}\n",
    "    for line in maf_block:\n",
    "        source, start, size, strand, srcSize, seq = line.split()[1:]\n",
    "        output[source] = {\n",
    "                'start': int(start),\n",
    "                'size': int(size),\n",
    "                'strand': strand,\n",
    "                'srcSize': int(srcSize),\n",
    "                'seq': seq\n",
    "                }\n",
    "    return output\n",
    "\n",
    "def extract_calls(maf_file,\n",
    "                  ref_species='hg19',\n",
    "                  contig='1',\n",
    "                  min_length=1):\n",
    "    \"\"\"\n",
    "    maf_file = bgziped .maf file (str)\n",
    "    ref_species = reference source for the .maf file (str)\n",
    "    contig = chromosome to extract allele calls for (str)\n",
    "    min_length = minimum alignment length required to extract calls (int)\n",
    "    ----------\n",
    "    Outputs a .txt file to stdout in the format of:\n",
    "    conting    pos    ref    target\n",
    "    \"\"\"\n",
    "    chromosome = 'chr'+contig\n",
    "    reference = ref_species+'.'+chromosome\n",
    "    infile = gzip.open(maf_file, 'rt')\n",
    "    # outfile = sys.stdout (Don't need this for testing.)\n",
    "    allele = str.maketrans('-acgtn', 'NACGTN')\n",
    "    blocks = maf_reader(infile)\n",
    "    maf_dicc = {}\n",
    "    for block in blocks:\n",
    "        block_info = maf_parser(block)\n",
    "        sequences = block_info.keys()\n",
    "        if reference != list(sequences)[0]:\n",
    "            continue\n",
    "        elif block_info[reference]['size'] < min_length:\n",
    "            continue\n",
    "        else:\n",
    "            true_len = block_info[reference]['size']\n",
    "            align_len = len(block_info[reference]['seq'])\n",
    "            ref_ind = [\n",
    "                    i for i in range(align_len)\n",
    "                    if block_info[reference]['seq'][i] != '-'\n",
    "                    ]\n",
    "            allele_calls = {}\n",
    "            positions = range(\n",
    "                    block_info[reference]['start']+1,\n",
    "                    block_info[reference]['start']+1 + true_len,\n",
    "                    )\n",
    "            allele_calls[reference] = block_info[reference]['seq'].\\\n",
    "                replace('-', '').translate(allele)\n",
    "            allele_calls[list(sequences)[1]] = block_info[list(sequences)[1]]['seq'].\\\n",
    "                translate(allele)\n",
    "            for j in range(true_len):\n",
    "                maf_dicc[positions[j]] = (\n",
    "                    allele_calls[list(sequences)[0]][j],\n",
    "                    allele_calls[list(sequences)[1]][ref_ind[j]],\n",
    "                )\n",
    "    return maf_dicc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3a761225",
   "metadata": {},
   "outputs": [],
   "source": [
    "### New Optimized Code ###\n",
    "\n",
    "\n",
    "# Define a function to extract the sequences in an alignment block from a MAF file.\n",
    "def maf_aln_block_generator(maf_file):\n",
    "    \"\"\"\n",
    "    Iterates through a MAF file and yields the sequence alignment blocks.\n",
    "    \"\"\"\n",
    "    # Intialize a list to store the sequences in the alignment block.\n",
    "    aln_block = []\n",
    "    # Iterate through every line.\n",
    "    for line in maf_file:\n",
    "        # If this is the start of a new alignment block.\n",
    "        if line.startswith('a'):\n",
    "            # If there are sequences from the previous alignment block.\n",
    "            if aln_block:\n",
    "                # Yield the sequences in the current alignment block.\n",
    "                yield aln_block\n",
    "                # Reset the list to store the sequences in the alignment block.\n",
    "                aln_block = []\n",
    "        # Else-if this a sequence within the alignment block.\n",
    "        elif line.startswith('s'):\n",
    "            # Update the alignment block.\n",
    "            aln_block.append(line.strip())\n",
    "    # If the last alignment block hasn't been processed.\n",
    "    if aln_block:\n",
    "        # Yield the sequences in the final alignment block.\n",
    "        yield aln_block\n",
    "\n",
    "# Define a function to compile the sequence fields for sequences in a MAF alignment block.\n",
    "def compile_seq_field_info(maf_aln_block):\n",
    "    \"\"\"\n",
    "    Extracts the sequence fields (see below) per sequence from a MAF alignment block.\n",
    "    \n",
    "    src: Name of the source sequence (alignment.chromosome).\n",
    "    start: Start of the aligning region in the source sequence (zero-based).\n",
    "    size: The length of the aligning region in the source sequence (non-dash characters only).\n",
    "    strand: If \"-\", then the alignment is to the reverse-complemented source.\n",
    "    srcSize: The size of the entire source sequence, not just the parts involved in the alignment.\n",
    "    seq: The nucleotides in the alignment including any insertions (\"-\").\n",
    "    \"\"\"\n",
    "    # Intialize a dictionary to store the sequence fields.\n",
    "    seq_fields = {}\n",
    "    # Iterate through every sequence in the alignment block.\n",
    "    for seq_line in maf_aln_block:\n",
    "        # Extract the sequence fields.\n",
    "        src, start, size, strand, srcSize, seq = seq_line.split()[1:]\n",
    "        # Update the sequence fields dictionary.\n",
    "        seq_fields[src] = {\n",
    "            'start': int(start), 'size': int(size),\n",
    "            'strand': strand, 'srcSize': int(srcSize), 'seq': seq,\n",
    "        }\n",
    "    return seq_fields\n",
    "\n",
    "# Define a function to build a MAF alignment dictionary.\n",
    "def build_maf_aln_dicc(maf_file, chrom, aln_block_len_thresh=1):\n",
    "    \"\"\"\n",
    "    Build a MAF alignment dictionary for a specified chromosome.\n",
    "    \n",
    "    maf_file: Gzipped .maf file.\n",
    "    chrom: Chromosome to extract allele calls for.\n",
    "    aln_block_len_thresh: Minimum alignment block length required to extract allele calls from.\n",
    "    \"\"\"\n",
    "    # Intialize the chromosome id.\n",
    "    chrom_id = f'chr{chrom}'\n",
    "    # Intialize the reference sequence source.\n",
    "    ref_seq_src = f'hg19.{chrom_id}'\n",
    "    # Translation table to convert bases.\n",
    "    vcf_trans = str.maketrans('-acgtn', 'NACGTN')\n",
    "    # Intialize a dictionary to store the MAF alignment.\n",
    "    maf_aln_dicc = {}\n",
    "    \n",
    "    # Open the MAF file.\n",
    "    with gzip.open(maf_file, 'rt') as infile:\n",
    "        # Iterate through every alignment block.\n",
    "        for aln_block in maf_aln_block_generator(infile):\n",
    "            # Extract the sequence fields.\n",
    "            seq_fields = compile_seq_field_info(aln_block)\n",
    "            # Grab the sequence fields for the refernce source.\n",
    "            ref_seq_fields = seq_fields.get(ref_seq_src)\n",
    "            \n",
    "            # If the first sequence is the reference sequence and passes the alignment block length threshold.\n",
    "            if (ref_seq_src == list(seq_fields.keys())[0]) and (seq_fields[ref_seq_src]['size'] >= aln_block_len_thresh):\n",
    "                # Grab the sequence fields for the non-refernce source.\n",
    "                alt_seq_fields = next((val for key, val in seq_fields.items() if key != ref_seq_src), None)\n",
    "                \n",
    "                # If the non-refernce source alignment is found.\n",
    "                if alt_seq_fields:\n",
    "                    # Grab the reference sequence.\n",
    "                    ref_seq = ref_seq_fields['seq']\n",
    "                    # Grab the length of the reference sequence (ie the number of ungapped bases).\n",
    "                    ref_seq_len = ref_seq_fields['size']\n",
    "                    # Intialize a generator for the non-gap indicies in the reference sequence alignment.\n",
    "                    nogap_idx_iter = iter(i for i, base in enumerate(ref_seq) if base != '-')\n",
    "                    # Generate the genomic positions.\n",
    "                    positions = range(ref_seq_fields['start'] + 1, ref_seq_fields['start'] + 1 + ref_seq_len)\n",
    "                    # Extract the the translated sequences.\n",
    "                    ref_seq_trans = ref_seq.replace('-', '').translate(vcf_trans)\n",
    "                    alt_seq_trans = alt_seq_fields['seq'].translate(vcf_trans)\n",
    "                    \n",
    "                    # Iterate through the reference sequence.\n",
    "                    for j, pos in enumerate(positions):\n",
    "                        # Update the dictionary.\n",
    "                        maf_aln_dicc[pos] = (ref_seq_trans[j], alt_seq_trans[next(nogap_idx_iter)])\n",
    "    return maf_aln_dicc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7c59824d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chr1: It is True that the legacy and new code produce the same results.\n",
      "Chr2: It is True that the legacy and new code produce the same results.\n",
      "Chr3: It is True that the legacy and new code produce the same results.\n",
      "Chr4: It is True that the legacy and new code produce the same results.\n",
      "Chr5: It is True that the legacy and new code produce the same results.\n",
      "Chr6: It is True that the legacy and new code produce the same results.\n",
      "Chr7: It is True that the legacy and new code produce the same results.\n",
      "Chr8: It is True that the legacy and new code produce the same results.\n",
      "Chr9: It is True that the legacy and new code produce the same results.\n",
      "Chr10: It is True that the legacy and new code produce the same results.\n",
      "Chr11: It is True that the legacy and new code produce the same results.\n",
      "Chr12: It is True that the legacy and new code produce the same results.\n",
      "Chr13: It is True that the legacy and new code produce the same results.\n",
      "Chr14: It is True that the legacy and new code produce the same results.\n",
      "Chr15: It is True that the legacy and new code produce the same results.\n",
      "Chr16: It is True that the legacy and new code produce the same results.\n",
      "Chr17: It is True that the legacy and new code produce the same results.\n",
      "Chr18: It is True that the legacy and new code produce the same results.\n",
      "Chr19: It is True that the legacy and new code produce the same results.\n",
      "Chr20: It is True that the legacy and new code produce the same results.\n",
      "Chr21: It is True that the legacy and new code produce the same results.\n",
      "Chr22: It is True that the legacy and new code produce the same results.\n",
      "ChrX: It is True that the legacy and new code produce the same results.\n",
      "ChrY: It is True that the legacy and new code produce the same results.\n"
     ]
    }
   ],
   "source": [
    "# For every chromosome.\n",
    "for chrom in [i for i in range(1, 23)]+['X', 'Y']:\n",
    "    # Update the dictionaries.\n",
    "    old_maf = extract_calls('../data/panTro/hg19.panTro6.synNet.maf.gz', 'hg19', str(chrom), 1)\n",
    "    new_maf = build_maf_aln_dicc('../data/panTro/hg19.panTro6.synNet.maf.gz', chrom)\n",
    "    # Sanity check.\n",
    "    old_equals_new = old_maf == new_maf\n",
    "    print(f'Chr{chrom}: It is {old_equals_new} that the legacy and new code produce the same results.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d99a0593",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "anc_der_intro_env",
   "language": "python",
   "name": "anc_der_intro_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
